File: Java/leetcode/editor/en/481.magical-string.java
import java.util.ArrayList;

//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
    public int magicalString(int n) {
        ArrayList<Integer> ans = new ArrayList<>();
        ans.add(1);
        boolean one = false;
        int counter = 1;
        int index = 1;
        int new_element;

        while (counter < n) {
            new_element = ans.get(ans.size()-1) == 1 ? 2 : 1;
            index ++;

            if (one) {
                counter ++;
                ans.add(new_element);
            } else {
                counter += 2;
                ans.add(new_element);
                ans.add(new_element);
            }

            one = ans.get(index) == 1 ? true : false;
        }
        int result = 0;
        for (int i=0; i<n; i++) {
            if (ans.get(i) == 1)
                result ++;
        }
        return  result;


    }
}
//leetcode submit region end(Prohibit modification and deletion)

File: 22FALL/481.magical-string.py
#
# @lc app=leetcode id=481 lang=python3
#
# [481] Magical String
#

# @lc code=start
class Solution:
    def magicalString(self, n: int) -> int:
        self.ans = [1]
        one = False
        counter = 1
        index = 1

        while counter < n:
            new_element = 1 if self.ans[-1] == 2 else 2
            index += 1
            if one:
                counter += 1
                self.ans.append(new_element)
            else:
                counter += 2
                self.ans.append(new_element)
                self.ans.append(new_element)

            if self.ans[index] == 1:
                one = True
            else:
                one = False
        answer = 0
        for i in range(n):
            if self.ans[i] == 1:
                answer += 1
        return answer

# @lc code=end

